### ISSUE TITLE
Reason for the STM32L072 when there is an ESP32 on board?

### ISSUE THREAD

**Issue by expaso** [kittengineering/Omgevingsmonitor_Hardware#2]
What is the reason for a separate ESP32-C3 and an STM32L072?

The ESP32 is much faster, has more flash, and more RAM. In this design, it's purely used as a WIFI chip, but will run user-code no problem, and even faster than the STM.

https://www.espressif.com/sites/default/files/documentation/esp32-c3-mini-1_datasheet_en.pdf

It looks to me that for a low-power device, you would avoid having multiple processors at work..
So..out of curiosity..  what made the design choice?

**Comment by tjallingk** [comment]
i think i had nothing to do with making it. but guess its lower power to have running continuously as the esp is in sleep most the time except when sending data so its on 1% or less of the time thus using even less power and esp32 is quite a power hungry chip for solar and a small battery

**Comment by expaso** [comment]
Thank you for the info!

The ESP does 5uA while in deep sleep with RTC timers enabled, and 1 uA when not enabled at all. Should be no problem.

The STM is indeed slower but consumes allmost nothing at all. 
According to the schematics, the STM has control over the ESPs flash functions, so I think we all know where this is heading for ..

We want this thing to run ESPHome (or at least MQTT) and connect to Home Assistant 😁..
And still run on solar power in the winter...

Totally doable! Challenge accepted 👍🏻😅

The STM is so low power, it won't make much of a difference anyway looking at the specs.

**Comment by neverstoppedaskingwhy** [comment]
@expaso I'm also interested in (additionally) providing the measurements via MQTT. The current JSON in the HTTP post contains quite a bit of overhead and could be slimmed down, but would already be usable as a MQTT payload. The implementation of the AT commands necessary for establishing the connection to the MQTT broker and publish the message, may not be too difficult.

**Comment by expaso** [comment]
@neverstoppedaskingwhy It's in the basis not that difficult.

In this hardwaredesign the STM seems to be the 'master' processor, but is fairly low power.
the ESP is just used as WIFI chip.

In my vision, we could use the STM to make all the measurements (sensors are also connected to the STM), drive the leds and periodically wake the ESP. The ESP could perfectly be running ESPHome or derived. The ESP would then do the heavy lifting of connecting to the WIFI, gather the measurements from the STM, postprocess, and send them to MQTT or HomeAssistant for that matter.

This way, the STM code is kept very slim and simple and is used as a PMU /  sensor-data aggregator and IO bridge only. 
Now we have paved the road for things like encryption, authentication, internal webserver, OTA firmware update of the ESP etc..

But I would love to hear other opinions as well..

**Comment by neverstoppedaskingwhy** [comment]
The state machine after HTTPCPOST waits for > before sending the message with the json. After sending the message, it expects an OK, which appears to be after an http 200 return code. 

When I run the Omgevingsmonitor against a local python http server, this server returns an error on http post instead. In this case the Omgevingsmonitor does not "sleep" for 5 minutes and posts new values ~ every 1 second. 

For now, I changed the firmware to use a hard-coded SSID and password, a simplified JSON with only the relevant data (114 byte in size), and use a local http url, with the ip address and port of my synology nas where node-red is running. I also reduced the sleep interval from 300 seconds to 120 seconds.

In node-red, I use a simple flow with an http-in node and an http response node. Also a mqtt out node, to publish the json payload as is. I created a simple node-red-dashboard 2.0 page to display the current values and a chart with the values of the last 48 hours.

I have the Omgevingsmonitor plugged into power. Today I noticed that the Omgevingsmonitor was no longer attempting to connect to wifi to send the data. After a restart everything worked fine. This was after about approx. 33 hours. I will continue to monitor this further, to see if this issue repeats itself.

I will have to look into obtaining some hardware for debugging, to gain the necessary insights into the state machine before making any changes for direct mqtt publishing, but I'm not sure which hardware I need exactly.
